### PR TITLE
Idea: Ride type descriptor

### PR DESCRIPTION
**Pull request by IntelOrca**
So now that we have individual source files for each type of ride, I thought it would be a good place to move all the constant data specific to a ride type in each source.

Currently a lot of the ride data is broken up into different structs and arrays, so this is essentially combining them all into one struct, `ride_type_desc`.

One problem though is the MSVC does not support `.field =` syntax in C++, which is a shame because its a nice way of, not only specifying required fields, but also annotating each value without using comments.

``` c
ride_type_desc RideTypeDescMerryGoRound = {
    .flags                      = RIDE_TYPE_DESC_FLAG_MUSIC_ON_DEFAULT |
                                  RIDE_TYPE_DESC_FLAG_ALLOW_MUSIC |
                                  RIDE_TYPE_DESC_FLAG_HAS_ENTRANCE_EXIT |
                                  RIDE_TYPE_DESC_FLAG_SINGLE_SESSION |
                                  RIDE_TYPE_DESC_FLAG_FLAG4_13,

    .operating_modes            = (1 << RIDE_MODE_ROTATION),
    .breakdowns                 = (1 << BREAKDOWN_SAFETY_CUT_OUT) | (1 << BREAKDOWN_CONTROL_FAILURE),

    .default_music_style        = MUSIC_STYLE_FAIRGROUND_ORGAN,

    .max_height                 = 12,
    .clearance_height           = 64,
    .z_offset                   = 3,
    .max_friction               = 255,
    .z                          = 2,
    .typical_price_multiplier   = 1,
    .bonus_value                = 45,

    .lift_hill_sound            = 255,
    .lift_hill_min_speed        = 5,
    .lift_hill_max_speed        = 5,

    // Prices
    .initial_upkeep_cost        = 50,
    .cost_per_track_piece       = 0,
    .price_primary              = 10,
    .price_secondary            = 0,
    .price_track                = 115,
    .price_supports             = 2,

    .value_rating_multiplier    = { 50, 10, 0 },

    // Strings
    .vehicle_name               = 1264,
    .structure_name             = 1278,
    .station_name               = 1257
};
```

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
@IntelOrca So the example above isn't possible? What about adding a build step which parses a YAML file and spits out a C file?

**Comment by Gymnasiast** [comment]
Good idea. We have so many separate tables for all the data.
If the `.field =` syntax doesn't work we can use comments instead. Possibly even inline if that increases readability.

File: src/ride/ride_type.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	rating_tuple value_rating_multiplier;
+
+	// Strings
+	rct_string_id vehicle_name;
+	rct_string_id structure_name;
+	rct_string_id station_name;
+} ride_type_desc;
+
+enum {
+	// All rides that have more than one piece than can be built for them,
```
Comment (Gymnasiast) [inline src/ride/ride_type.h]: Should be "one piece tha**t** can be built"

**Comment by IntelOrca** [comment]
@marijnvdwerf hmm it would be good avoid external preprocessing if necessary, but we will most likely need to generate the string ids anyway, so its not out the question.

**Comment by janisozaur** [comment]
Is the `.field =` syntax still not supported in VS15?

**Comment by IntelOrca** [comment]
@janisozaur is it not part of the C++ spec. its a C99 feature but not a C++ feature. But GCC obviously allow it to work anyway where as MSVC do not. Its very annoying.

**Comment by Gymnasiast** [comment]
Any news on this? I still find it a very good idea.

**Comment by IntelOrca** [comment]
Need volunteers to group the data together for all the ride types.

**Comment by Broxzier** [comment]
Instead of the `.field=` syntax, we can write it fully like `RideTypeDescMerryGoRound.field=`

**Comment by IntelOrca** [comment]
@Broxzier how?

**Comment by camthesaxman** [comment]
Why not just use C for this if you want designated initializers?

**Comment by IntelOrca** [comment]
Another solution:
```
static constexpr ride_type_desc GetRideTypeDesc()
{
  ride_type_desc desc = { 0 };
  desc.default_music_style        = 0;
  desc.max_height                 = 12;
  desc.clearance_height           = 64;
  desc.z_offset                   = 3;
  desc.max_friction               = 255;
  desc.z                          = 2;
  desc.typical_price_multiplier   = 1;
  desc.bonus_value                = 45;
  return desc;
}
const ride_type_desc RideTypeDescMerryGoRound = GetRideTypeDesc();
```

For gcc and clang, this will define static data like we want. However for MSVC, it still sets the field at startup (using constexpr on the field crashes the compiler).

**Comment by camthesaxman** [comment]
Do we have to use C++ for this? I don't see any problem with defining the data in a .c file and putting the code in a .cpp file.

**Pull request closed without merging**